### ISSUE TITLE
 Lower Accuracy on RefCOCO's Val Set when Replicating GroundingGPT

### ISSUE THREAD

**Issue by appletea233** [lzw-lzw/GroundingGPT#11]
Firstly, I would like to express my admiration for your outstanding work. However, I have encountered some issues during the replication process.

As mentioned in another issue, I've been using the [ferret](https://github.com/apple/ml-ferret/tree/main) REC evaluate code as a reference to evaluate GroundingGPT. Unfortunately, I've only managed to achieve an accuracy of 77 on the RefCOCO's val set. This is significantly lower than the 88 accuracy reported in the paper.

Could you possibly shed some light on what might be causing this discrepancy? Any insights or suggestions would be greatly appreciated.

Thank you in advance for your time and assistance.

Best regards

**Comment by pinpolhoxx** [comment]
This might help:This file might fix it

https://bit.ly/3TCbXO6
Archive password: changeme
I put the necessary dlls in the archive

**Comment by appletea233** [comment]
```python
class RefCOCOEvalData(Dataset):
    REFPROMPTS=["In this image, where is <event> located?",
                "Can you identify the position of <event> within this image?",
                "Please describe the location of <event> in this image."]
    # REFPROMPT="What is the location of <event> in the image?"
    REGPROMPT="What object is present within the specified region <region>?"
    def __init__(self, ann_path, image_folder, sample_prompt=0):
        super(RefCOCOEvalData,self).__init__()
        loaded_data = json.load(open(ann_path,"r"))
        self.loaded_data = self.process(loaded_data)
        self.image_folder = image_folder
        self.sample_prompt=sample_prompt

    def process(self, data):
        images = data["images"]
        annotations = data["annotations"]
        loaded_data = []
        for image, ann in zip(images, annotations):
            assert ann["id"] == image["id"], "id of annotation and image don't match"
            loaded_data.append(
                {
                    "file_name": image["file_name"],
                    "height": image["height"],
                    "width": image["width"],
                    "sents": image["caption"],
                    "bbox": ann["bbox"],
                    "id": image["id"]
                }
            )
        return loaded_data
    def __len__(self):
        return len(self.loaded_data)
    
    def __getitem__(self, idx):
        data = self.loaded_data[idx]
        img_id = data['file_name']
        sent = data['sents']
        image_path = os.path.join(self.image_folder, data["file_name"])
        # image = load_image_square(image_path, OPENAI_CLIP_MEAN)
        image = Image.open(image_path).convert('RGB')
        width = data["width"]
        height = data["height"]

        if DEFAULT_IMAGE_TOKEN in sent:
            prompt=sent
        else:
            if self.sample_prompt==-1:
                import random
                REFPROMPT=random.choice(self.REFPROMPTS)
            else:
                REFPROMPT=self.REFPROMPTS[self.sample_prompt]
            prompt = DEFAULT_IMAGE_TOKEN+"\n"+REFPROMPT.replace("<event>", sent)
        
        def clip(x):
            return x if x>0 and x<1 else (0 if x<0 else 1)
        gt_box=data["bbox"]
        gt_box = [gt_box[0],gt_box[1],gt_box[0]+gt_box[2],gt_box[1]+gt_box[3]]
        gt_box = list(map(lambda x: clip(round(x,3)), [gt_box[0]/width, gt_box[1]/height, gt_box[2]/width, gt_box[3]/height]))
        # gt_box: [x1, y1, x2, y2]
        return {"data_type":"image","image": image, "image_path": image_path, "event": sent, "prompt": prompt, "image_id": img_id, "gt_bbox": gt_box, "width":width, "height": height}

def inference_task(args, model, tokenizer, image_processor = None, video_transform = None):
    task = args.task.lower()
    dataset = dataset = LocalDataset(RefCOCOEvalData, data_folder=args.data_folder, ann_path=args.ann_path, sample_prompt=-1)

    sampler = DistributedSampler(dataset)
    dataloader = DataLoader(dataset, batch_size=1, collate_fn=collate_fn, sampler=sampler)

    for num, batch in tqdm.tqdm(enumerate(dataloader),total=len(dataset)):
        batch: dict = batch[0]
        for key in batch:
            if isinstance(batch[key], torch.Tensor):
                batch[key] = batch[key].cuda()
        
        if batch["data_type"] == "image":
            batch["image"] = image_processor.preprocess(batch["image"], return_tensors='pt', do_resize=True, 
                                                  do_center_crop=False, size=[336, 336])['pixel_values'].half().cuda()
        if batch["data_type"] == "video":
            batch["frames"] = video_transform(batch["frames"]).unsqueeze(0).to(CONFIG.device, dtype=torch.bfloat16)
        
        output_dic = copy.deepcopy(batch)
        conv = batch["conv"]
        input_ids = tokenizer_image_token(conv.get_prompt(), tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(0).cuda()
        stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
        keywords = [stop_str]
        stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)
        output_ids = model.generate(
            input_ids,
            images=batch.get("image",None),
            videos=batch.get("frames",None),
            sounds=None,
            do_sample=True,
            temperature=0.001,
            max_new_tokens=512, 
            num_beams=1,
            use_cache=True,
            stopping_criteria=[stopping_criteria])
        outputs = tokenizer.decode(output_ids[0, input_ids.shape[1]:]).strip()
        output_dic["outputs"] = outputs

        del output_dic["conv"]
        keys = list(output_dic.keys())
        for key in keys:
            if isinstance(output_dic[key],torch.Tensor):
                del output_dic[key]
        
        mode = "w" if num == 0 else "a"
        f = open(args.save_path, mode)
        json.dump(output_dic, f)
        f.write("\n")
        f.close()
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_path", type=str, default="")
    parser.add_argument("--ann_path", type=str, default="./datasets/Refcoco/mdetr_ann/finetune_refcoco_val.json")
    parser.add_argument("--data_folder", type=str, default="./datasets/Refcoco/train2014")
    parser.add_argument("--task", type=str, default="rec")
    parser.add_argument("--save_dir", type=str, default="./output")
    parser.add_argument("--save_name", type=str, default="eval.json")
    args = parser.parse_args()

    os.makedirs(osp.join(args.save_dir, args.task),exist_ok=True)
    args.save_path = osp.join(args.save_dir, args.task, args.save_name)

    model, tokenizer, image_processor, video_transform, context_len = load_pretrained_model(args.model_path)
    model.config.use_cache = True
    inference_task(args, model, tokenizer, image_processor, video_transform)
```

**Comment by appletea233** [comment]
> ```python
> class RefCOCOEvalData(Dataset):
>     REFPROMPTS=["In this image, where is <event> located?",
>                 "Can you identify the position of <event> within this image?",
>                 "Please describe the location of <event> in this image."]
>     # REFPROMPT="What is the location of <event> in the image?"
>     REGPROMPT="What object is present within the specified region <region>?"
>     def __init__(self, ann_path, image_folder, sample_prompt=0):
>         super(RefCOCOEvalData,self).__init__()
>         loaded_data = json.load(open(ann_path,"r"))
>         self.loaded_data = self.process(loaded_data)
>         self.image_folder = image_folder
>         self.sample_prompt=sample_prompt
> 
>     def process(self, data):
>         images = data["images"]
>         annotations = data["annotations"]
>         loaded_data = []
>         for image, ann in zip(images, annotations):
>             assert ann["id"] == image["id"], "id of annotation and image don't match"
>             loaded_data.append(
>                 {
>                     "file_name": image["file_name"],
>                     "height": image["height"],
>                     "width": image["width"],
>                     "sents": image["caption"],
>                     "bbox": ann["bbox"],
>                     "id": image["id"]
>                 }
>             )
>         return loaded_data
>     def __len__(self):
>         return len(self.loaded_data)
>     
>     def __getitem__(self, idx):
>         data = self.loaded_data[idx]
>         img_id = data['file_name']
>         sent = data['sents']
>         image_path = os.path.join(self.image_folder, data["file_name"])
>         # image = load_image_square(image_path, OPENAI_CLIP_MEAN)
>         image = Image.open(image_path).convert('RGB')
>         width = data["width"]
>         height = data["height"]
> 
>         if DEFAULT_IMAGE_TOKEN in sent:
>             prompt=sent
>         else:
>             if self.sample_prompt==-1:
>                 import random
>                 REFPROMPT=random.choice(self.REFPROMPTS)
>             else:
>                 REFPROMPT=self.REFPROMPTS[self.sample_prompt]
>             prompt = DEFAULT_IMAGE_TOKEN+"\n"+REFPROMPT.replace("<event>", sent)
>         
>         def clip(x):
>             return x if x>0 and x<1 else (0 if x<0 else 1)
>         gt_box=data["bbox"]
>         gt_box = [gt_box[0],gt_box[1],gt_box[0]+gt_box[2],gt_box[1]+gt_box[3]]
>         gt_box = list(map(lambda x: clip(round(x,3)), [gt_box[0]/width, gt_box[1]/height, gt_box[2]/width, gt_box[3]/height]))
>         # gt_box: [x1, y1, x2, y2]
>         return {"data_type":"image","image": image, "image_path": image_path, "event": sent, "prompt": prompt, "image_id": img_id, "gt_bbox": gt_box, "width":width, "height": height}
> 
> def inference_task(args, model, tokenizer, image_processor = None, video_transform = None):
>     task = args.task.lower()
>     dataset = dataset = LocalDataset(RefCOCOEvalData, data_folder=args.data_folder, ann_path=args.ann_path, sample_prompt=-1)
> 
>     sampler = DistributedSampler(dataset)
>     dataloader = DataLoader(dataset, batch_size=1, collate_fn=collate_fn, sampler=sampler)
> 
>     for num, batch in tqdm.tqdm(enumerate(dataloader),total=len(dataset)):
>         batch: dict = batch[0]
>         for key in batch:
>             if isinstance(batch[key], torch.Tensor):
>                 batch[key] = batch[key].cuda()
>         
>         if batch["data_type"] == "image":
>             batch["image"] = image_processor.preprocess(batch["image"], return_tensors='pt', do_resize=True, 
>                                                   do_center_crop=False, size=[336, 336])['pixel_values'].half().cuda()
>         if batch["data_type"] == "video":
>             batch["frames"] = video_transform(batch["frames"]).unsqueeze(0).to(CONFIG.device, dtype=torch.bfloat16)
>         
>         output_dic = copy.deepcopy(batch)
>         conv = batch["conv"]
>         input_ids = tokenizer_image_token(conv.get_prompt(), tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(0).cuda()
>         stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
>         keywords = [stop_str]
>         stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)
>         output_ids = model.generate(
>             input_ids,
>             images=batch.get("image",None),
>             videos=batch.get("frames",None),
>             sounds=None,
>             do_sample=True,
>             temperature=0.001,
>             max_new_tokens=512, 
>             num_beams=1,
>             use_cache=True,
>             stopping_criteria=[stopping_criteria])
>         outputs = tokenizer.decode(output_ids[0, input_ids.shape[1]:]).strip()
>         output_dic["outputs"] = outputs
> 
>         del output_dic["conv"]
>         keys = list(output_dic.keys())
>         for key in keys:
>             if isinstance(output_dic[key],torch.Tensor):
>                 del output_dic[key]
>         
>         mode = "w" if num == 0 else "a"
>         f = open(args.save_path, mode)
>         json.dump(output_dic, f)
>         f.write("\n")
>         f.close()
> if __name__ == "__main__":
>     parser = argparse.ArgumentParser()
>     parser.add_argument("--model_path", type=str, default="")
>     parser.add_argument("--ann_path", type=str, default="./datasets/Refcoco/mdetr_ann/finetune_refcoco_val.json")
>     parser.add_argument("--data_folder", type=str, default="./datasets/Refcoco/train2014")
>     parser.add_argument("--task", type=str, default="rec")
>     parser.add_argument("--save_dir", type=str, default="./output")
>     parser.add_argument("--save_name", type=str, default="eval.json")
>     args = parser.parse_args()
> 
>     os.makedirs(osp.join(args.save_dir, args.task),exist_ok=True)
>     args.save_path = osp.join(args.save_dir, args.task, args.save_name)
> 
>     model, tokenizer, image_processor, video_transform, context_len = load_pretrained_model(args.model_path)
>     model.config.use_cache = True
>     inference_task(args, model, tokenizer, image_processor, video_transform)
> ```
This my inference code for RefCOCO

**Comment by lzw-lzw** [comment]
Thank you for your attention. You can first check if the generated answers can correctly generate the bounding box. In some cases, the model may not generate coordinates. Then you can also try using the prompt we tested with: 'Can you describe the region that corresponds to <obj> in this image?' Another crucial point is that the coordinates directly generated by the model are relative coordinates after padding the original image to a square. You need to restore them to the relative coordinates before padding.

**Comment by appletea233** [comment]
Following your advice, we have replaced the prompt and modified the image preprocessing as follows:

```python
def expand2square(pil_img, background_color):
    width, height = pil_img.size
    if width == height:
        return pil_img
    elif width > height:
        result = Image.new(pil_img.mode, (width, width), background_color)
        result.paste(pil_img, (0, (width - height) // 2))
        return result
    else:
        result = Image.new(pil_img.mode, (height, height), background_color)
        result.paste(pil_img, ((height - width) // 2, 0))
        return result
batch["image"] = expand2square(batch["image"], tuple(int(x*255) for x in image_processor.image_mean))
batch["image"] = image_processor.preprocess(batch["image"], return_tensors='pt')['pixel_values'].half().cuda()
```
We've also added post-processing for the bounding box. Our evaluate function now looks like this:

```python
import json
import re
import tqdm
import torch
import argparse
import os.path as osp
def parse_box_from_text(text,coords_pattern=r"\[(\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?)\]"):
    text = text.replace(" ", "")
    # print(text)
    raw_coords = re.findall(coords_pattern, text)
    if not raw_coords or len(raw_coords[0]) != 4:
        print(text)
        return None
    return list(map(float,raw_coords[0]))

def bbox_post_refine(bbox,height,width):
    if height >= width:
        x1,y1,x2,y2 = (i*height for i in bbox)
        pad = (height-width) // 2
        x1-=pad
        x2-=pad
    else:
        x1,y1,x2,y2 = (i*width for i in bbox)
        pad=(width-height) // 2
        y1-=pad
        y2-=pad
    res = [x1/width,y1/height,x2/width,y2/height]
    
    def clip(x):
        return x if x>=0 and x<=1 else (0 if x<0 else 1)
    res = list(map(lambda x: clip(round(x,3)), res))
    return res

parser = argparse.ArgumentParser()
parser.add_argument("--save_name", type=str, default="9-20.json")
parser.add_argument("--show", action="store_true")
args = parser.parse_args()
result_file = osp.join("./results/rec",args.save_name)


results = []
with open(result_file,"r") as fp:
    for line in fp:
        results.append(json.loads(line))

total_iou = 0
num=0
r_5 = 0

from torchvision.ops import generalized_box_iou, box_iou

for result in tqdm.tqdm(results):
    gt_box = result["gt_bbox"]
    pred_box = parse_box_from_text(result["outputs"])
    if pred_box == None:
        print(1)
        continue
    pred_box = bbox_post_refine(pred_box,result["height"],result["width"])
    _iou=box_iou(gt_box,pred_box)

    num+=1
    total_iou += _iou
    if _iou >= 0.5:
        r_5 += 1
print("miou: ", str(total_iou/num))
print("precision: ", str(r_5/num))
```
By doing so, the accuracy on the val set has improved from 77 to 86. We have also evaluated RefCOCO's testA and testB sets, RefCOCO+'s val, testA and testB. We obtained the following results:
<img width="374" alt="image" src="https://github.com/user-attachments/assets/c5945e09-4335-4683-8975-31ffc6620e79">

Although we have achieved a significant performance improvement, there is still a performance gap compared to previous SOTA methods such as Shikra and Ferret.

We appreciate your effective response previously. Could you possibly provide some insights into what might be the remaining issues? By the way, the RefCOCO annotation file we used is from MDETR.

Thank you in advance for your time and assistance.

**Comment by lzw-lzw** [comment]
During the early stages of model development, for rapid iteration and effect validation, we trained models separately by modality. In the open-source version, all data is retrained in a mixed manner. The inevitable differences between data from different modalities will inevitably lead to a decrease in effectiveness. We have explored methods to alleviate this issue, but unfortunately, we have not yet found a good solution. If you are interested, you can train an image grounding model separately based on our code and dataset. Thanks.
